### PR TITLE
start admin interface earlier

### PR DESCRIPTION
**Pull request by stevenctl**
In some scenarios, it would be very useful to call debug functions before the server is ready to debug startup issues.

### CODE REVIEW

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -199,6 +199,24 @@ func NewServer(args *PilotArgs, initFuncs ...func(*Server)) (*Server, error) {
...
+			return wh.Config.RawTemplates
+		}
+		return map[string]string{}
+	}
+	if err := s.initIstiodAdminServer(args, whc); err != nil {
+		return nil, fmt.Errorf("error initializing debug server: %v", err)
+	}
+
 	// Apply custom initialization functions.
 	for _, fn := range initFuncs {
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/server.go:221]: seems like we should at least apply these first -they could change the args?

File: pilot/pkg/bootstrap/server.go [hunk 2]
```diff
@@ -199,6 +199,24 @@ func NewServer(args *PilotArgs, initFuncs ...func(*Server)) (*Server, error) {
...
+	)
+	whc := func() map[string]string {
+		whMu.RLock()
+		defer whMu.RUnlock()
+		if wh != nil {
+			return wh.Config.RawTemplates
+		}
+		return map[string]string{}
+	}
+	if err := s.initIstiodAdminServer(args, whc); err != nil {
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/server.go:216]: do we know that we won't get crashes, etc from all the endpoints? For example kubeClient being nil may be problematic ?

Might be worth testing that things work or fail gracefully if we were to sleep(infinity) right after this

**Review by howardjohn — COMMENTED** [review]
overall lgtm

**Comment by hzxuzhonghu** [comment]
could be useful, make sense to me

**Comment by stevenctl** [comment]
had to actually move where we start serving on the http port.. this means we have to say no to readiness before that. also, tested every endpoint and guarded a few nil-pointer accesses. Even when I did hit them, they didn't bubble up and crash the app.

**Comment by stevenctl** [comment]
cc @howardjohn  @shankgan

**Review by shankgan — APPROVED** [review]

File: pilot/pkg/xds/debug.go [hunk 3]
```diff
@@ -1039,6 +1049,9 @@ func (s *DiscoveryServer) instancesz(w http.ResponseWriter, req *http.Request) {
 }
 
 func (s *DiscoveryServer) networkz(w http.ResponseWriter, req *http.Request) {
+	if s.Env == nil || s.Env.NetworkManager == nil {
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/debug.go:1052]: Would this introduce race for test

File: pilot/pkg/bootstrap/server.go [hunk 4]
```diff
@@ -209,6 +224,17 @@ func NewServer(args *PilotArgs, initFuncs ...func(*Server)) (*Server, error) {
...
 	prometheus.EnableHandlingTimeHistogram()
 
+	// make sure we have a readiness probe before serving HTTP to avoid marking ready too soon
+	s.addReadinessProbe("discovery", func() (bool, error) {
+		return s.XDSServer.IsServerReady(), nil
+	})
+	if err := s.initIstiodAdminServer(args, whc); err != nil {
+		return nil, fmt.Errorf("error initializing debug server: %v", err)
+	}
+	if err := s.serveHTTP(); err != nil {
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go:234]: Check http server is well shutdown in ut

**Review by howardjohn — APPROVED** [review]
